contrib/babelfishpg_common/src/datetime.c

```diff
@@ -663,6 +666,454 @@ datetime_pl_datetime(PG_FUNCTION_ARGS)
+ * {"DD MON YYYY", "DD YYYY MON", "MON DD YYYY", "MON YYYY DD", "YYYY MON DD", "YYYY DD MON"}
+ * where MON is valid month in text format then returns true if present.
+ * Additional to that if Month is present in 1st or 3rd field then will swap the field with 2nd
+ * field i.e., if the fields are of format {"DD YYYY MON", "MON DD YYYY", "MON YYYY DD", "YYYY DD MON"}
+ * then will be converted to {"DD MON YYYY", "YYYY MON DD"} because later the fields will be changed
+ * to format {"DD-MON-YYYY", "DD-YYYY-MON", "MON-DD-YYYY", "MON-YYYY-DD", "YYYY-MON-DD", "YYYY-DD-MON"}
+ * out of which only formats {"DD-MON-YYYY", "YYYY-MON-DD"} are valid.
+ */
+static bool containsInTextMonthFormat(int *ftype, char **field)
+{
+	int count_number = 0, count_string = 0;
+
+	#define SWAP_FIELDS(i, j) \
+		char* temp_field;	\
+		int temp_ftype;	\
+		temp_field = field[i];	\
+		temp_ftype = ftype[i];	\
+		field[i] = field[j];	\
+		field[j] = temp_field;	\
+		ftype[i] = ftype[j];	\
+		ftype[j] = temp_ftype;	\
+
+	for (int i = 0; i < 3; i++)
+	{
+		if (ftype[i] == DTK_NUMBER)
+			count_number++;
+		else if (ftype[i] == DTK_STRING)
+		{
+			/* Check whether the string is valid text month */
+			if (!isTextMonthPresent(field[i]))
+				return false;
+			count_string++;
+		}
+		else
+			return false;
+	}
+	if (count_number == 2 && count_string == 1)
+	{
+		/*
+		 * If the first field and third field is an string then swap with the second field
+		 * as when the date is given separatly then all different forms of
+		 * dates is supported. To avoid the conversion failure from `isTextMonthPresent`
+		 * later we are swapping earlier. And convert to Supported date formats which are
+		 * {"DD-MON-YYYY", "YYYY-MON-DD"}
+		 *
+		 * For example convert the input from "JULY 23 2000" to "23 JULY 2000"
+		 * and convert the input from "23 2000 JULY" to "23 JULY 2000".
+		 */
+		if (ftype[0] == DTK_STRING)
+		{
+			SWAP_FIELDS(0, 1);
+		}
+		else if (ftype[2] == DTK_STRING)
+		{
+			SWAP_FIELDS(1, 2);
+		}
+		return true;
+	}
+	return false;
+}
+
+/*
+ * pltsql_time_in
+ *		The function modifies the given input into valid form and
+ *		stores the output to result. If there are any errors then
+ *		an error message will be thrown.
+ * Returns
+ * 		false - If the sql_dialect is not an TSQL and proceed with
+ * 			the original behaviour
+ * 		true - If the sql_dialect is an TSQL
+ */
+bool pltsql_time_in(const char* str, int32 typmod, TimeADT *result)
+{
+	fsec_t fsec;
+	struct pg_tm tt, *tm = &tt;
+	int tz, nf, dterr, dtype, ftype[MAXDATEFIELDS];
+	char workbuf[MAXDATELEN + 1], *field[MAXDATEFIELDS];
+	StringInfo res;
+	regex_t time_regex;
+	char *pattern, *temp_field;
+	int len_str = strlen(str);
+	char *modified_str = (char*) palloc(len_str + 1);
+	int j = 0, i = 0;
+
+	/* If sql_dialect is not an TSQL then return false */
+	if (sql_dialect != SQL_DIALECT_TSQL)
+		return false;
+
+	/* Throw a common error message while casting to time datatype */
+	#define TIME_IN_ERROR()	\
```

Comment:
have we made sure that we have a test case for each code path which can throw a time_in_error in this function?

Reply:
Added tests in the JDBC test file, where we get TIME_IN_ERROR() cases i.e., negative cases for each type.